Pcap++/header/PcapLiveDeviceList.h

```diff
@@ -56,7 +68,9 @@ namespace pcpp
 
 		/**
 		 * @return A vector containing pointers to all live devices currently installed on the machine
+		 * @deprecated This method has been deprecated in favor of begin/end iteration API.
 		 */
+		PCPP_DEPRECATED("Deprecated in favor of begin/end iteration API")
 		const std::vector<PcapLiveDevice*>& getPcapLiveDevicesList() const { return m_LiveDeviceListView; };
```

Comment:
There are things you can do with a vector that you can't do with an iterator, such as getting the element at index X, for example: `PcapLiveDeviceList::getInstance().getPcapLiveDevicesList()[1]`, or get the first or last element. 

Of course we can implement all of those as part of `PcapLiveDeviceList`, but isn't it easier to just expose the vector? 🤔 

And if we expose the vector it's probably a bit confusing to also expose an iterator over the `PcapLiveDeviceList` class itself

Reply:
The argument about indexing is true, tho I don't really agree that exposing the vector is theright choice. It might be easier to do, but it will cause more headaches in the long run. 

One of which we already encountered when refactoring the pointer management to be automatic w.r.t the current class. That issue being the nessesity to have a vector alive (the "view" class) to return a reference to. This is the primary reason I have for this deprecation.

Also while I already mentioned it, conceptually, it does not look right to me to have a class named 'List' have to be queried for another container. If it was called 'Manager' or something like that, then fine, I guess. IMO 'List' implies a container and at the moment, what we have does not behave like a container.

Reply:
> The argument about indexing is true, tho I don't really agree that exposing the vector is the right choice. It might be easier to do, but it will cause more headaches in the long run.

Which headaches do you think it'll cause in the long run?



> One of which we already encountered when refactoring the pointer management to be automatic w.r.t the current class. That issue being the nessesity to have a vector alive (the "view" class) to return a reference to. This is the primary reason I have for this deprecation.

I agree that keeping the view isn't amazing, but it's good enough I think, so we can return the vector to the user if needed



> Also while I already mentioned it, conceptually, it does not look right to me to have a class named 'List' have to be queried for another container. If it was called 'Manager' or something like that, then fine, I guess. IMO 'List' implies a container and at the moment, what we have does not behave like a container.

I agree the name for this class could be improved, maybe "Manager" or "Container" would be better. But since those classes exist for a very long time changing them would be a pain

Reply:
> Which headaches do you think it'll cause in the long run?

Well, primarily the one that it binds the internal implementation to the external API, which makes potential changes to the implementation difficult, which we already encountered. I especially don't like it in these classes as they are one of the main blocks of the library and the API should be relatively stable.

> I agree that keeping the view isn't amazing, but it's good enough I think, so we can return the vector to the user if needed

Eh, I guess, I mostly added it as a temporary band-aid while I run though the deprecations... which I guess I should have known better how all temporary solutions and up...

One thing I kinda want is to unify the iteration and access API for all the different lists. At the moment it is kinda all over the place. `LiveDevice`, `PfRing` and `Dpdk` all expose a reference to vector, `RemoteDevice` exposes iterators and `Kni` exposes nothing. One of my ideas is to have a base templated 'List' container which implements all the accessors and then all the lists inherit from that.

> I agree the name for this class could be improved, maybe "Manager" or "Container" would be better. But since those classes exist for a very long time changing them would be a pain.

Eh, `List` is fine as a name, it is just that it would be better if it behaved like a proper container since it is already named like one.

Reply:
Assuming we go with your suggestion, we probably need to implement most of the vector getters like: count, first, last, getByIndex, iterator, etc. right?

Reply:
I think these ones would be sufficient:
- at / operator[]
- front / back
- (c)begin/(c)end
- maybe r(c)begin/r(c)end
- empty/size

These are the main ones that would most likely be used.

Reply:
It's a bit weird, because all of these method will call `m_LiveDeviceList` with the same method, it'll look like:
```cpp
methodX()
{
    return m_LiveDeviceList.methodX();
}
```

The only thing we gain from not exposing the vector is:
- We make `PcapLiveDeviceList` the container
- We make it read-only, so no one can modify it

Reply:
> n alternative could be to just return the result of `getAll*Devices()` by value instead of reference. That would also fix the issue of the internal implementation being bound to the API. At the expense of the list being generated every time it is called and not being updated after that (due to `reset()`, etc calls). Although I am hesitant to change a reference return to a value return directly as it might affect existing code.

If you think that's better then I'm ok with it. It will change the existing interface, but do you see any potential usage of this method that might break? 🤔

Reply:
> > n alternative could be to just return the result of `getAll*Devices()` by value instead of reference. That would also fix the issue of the internal implementation being bound to the API. At the expense of the list being generated every time it is called and not being updated after that (due to `reset()`, etc calls). Although I am hesitant to change a reference return to a value return directly as it might affect existing code.
> 
> If you think that's better then I'm ok with it. It will change the existing interface, but do you see any potential usage of this method that might break? 🤔

Well, I kinda threw it there as a potential alternative. Tho looking into it, the change from a reference to a value return looks like a minefield. 

See this code for example.
```
struct Object
{
	std::vector<int> const& getValueByRef() { return m_value; }
	std::vector<int> const getValue() { return m_value; }
	
	std::vector<int> m_value;
};

int main()
{ 
	Object obj;
        
        // This a fine and possibly common usage of the current implementation.
	std::vector<int> const& valueRef = obj.getValueByRef();
        // This is what the code might look like after a direct change.
        // It returns a temporary which's address is assigned to the reference, which compiles but is UB.
        // How long does the temporary live? Nobody knows, so the reference can become dead at any moment.
	std::vector<int> const& valueRef2 = obj.getValue(); 
	return 0;
}
```

So yeah... the return by value would require its own separate function to not break a lot of stuff. :/

On the other hand... tried to do the original idea of having public access methods similar to vector and I think it went pretty well, especially with the base class.